In [4]:
# 파일 불러오기

import numpy as np
import pandas as pd

score = pd.read_csv('./dataset/score.csv')
user = pd.read_csv('./dataset/user.csv')


In [5]:
print(score.info())
print(score.shape)
score.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   sno     14 non-null     int64  
 1   kor     12 non-null     float64
 2   eng     11 non-null     float64
 3   math    14 non-null     int64  
 4   year    14 non-null     int64  
dtypes: float64(2), int64(3)
memory usage: 688.0 bytes
None
(14, 5)


sno   kor   eng  math  year
0  1001  88.0  84.0    92  2020
1  1001   NaN  79.0    82  2021
2  1002  84.0  88.0    94  2020
3  1002  74.0   NaN    79  2021
4  1003  92.0  94.0    90  2020

In [6]:
print(user.info())
print(user.shape)
user.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   sno     7 non-null      int64 
 1   ename   7 non-null      object
 2   age     7 non-null      int64 
 3   gen     7 non-null      object
dtypes: int64(2), object(2)
memory usage: 352.0+ bytes
None
(7, 4)


sno ename  age gen
0  1001   송강호   33   m
1  1002   황정민   29   m
2  1003   김윤석   31   m
3  1004   설경구   27   m
4  1005   김혜수   32   f

<pre>
결측 있나?
object 있나?
데이터 읽으면 데이터 관계 파악★ -> sno로 이을 수 있겠다. score는 학생 당 2개년치 있어서 14행, 학생은 7행

### 1. 각 과목별 결측 데이터 수

In [7]:
# score.columns
# score.isna().sum() # --> 이게 정석(추천)
score[['kor', 'eng', 'math']].isna().sum()

kor     2
eng     3
math    0
dtype: int64

### 2. 2020년도 각 과목별 평균 점수

In [8]:
# score[컬럼][조건]
# score[조건][컬럼]

score[score['year']==2020].groupby('year')[['kor', 'eng', 'math']].mean()
## 가운데 .groupby 안 해도 되네 (220111 설명 하심)

kor        eng       math
year                                 
2020  88.857143  88.285714  91.428571

* 3. 성적이 없는 학생의 점수는 '50점'               로 채우기
* 3. 성적이 없는 학생의 점수는 '해당 과목의 최저 점수'로 채우기

In [9]:
# score50=score.fillna(50).copy()
# score50

In [10]:
# 강사님 -> 전체에 넣어버리는 것보다는 한 줄 씩 입력하는 게 안정적이긴 하다.

# col_list=['kor','eng'] # -> 근데 컬럼 이름 일일히 입력하기 귀찮음.
# for col in col_list:
#     score50[col].fillna(50, inplace=True)

In [11]:
# 전체 데이터프레임에서 결측 있는 컬럼만 가져오기 (시도) (문제를 잘못 이해함 ^^;;;;)

# for col in score.columns.tolist():
#     for row in score.index:
#         if score[row,col].isna()==True:
#             score50=score[row,col].fillna(50)

# for col in score.column.tolist():
#     for val in score.values:
#             if score[row,col].isna()==True:
#                 score50=score[row,col].fillna(50)
                
# for col in score.columns.tolist():
#     for row in score.values().tolist():
#         print(score[row,col])



In [12]:
#  근데 컬럼 이름 일일히 입력하기 귀찮음.
# 결측이 있는 컬럼만 가져오기
# df [ [df > 0] ] : 이용

score.isna().sum()[score.isna().sum()>0] # 시리즈로 나온다.
score.isna().sum()[score.isna().sum()>0].index.tolist()

['kor', 'eng']

In [13]:
#  근데 컬럼 이름 일일히 입력하기 귀찮음.
# 결측이 있는 컬럼만 가져오기
# by 최상현

print([i for i in score.columns if score[i].isna().sum()!=0])

# print([i for i in score.columns if score[i].isna().sum()>0]) (강사님)

['kor', 'eng']


In [14]:
# 틀렸다가 다시 푸니까 맞았는데 score.min() 완벽 이해는 못 함. -> 밑에 혼자 공부 참고

smin=score[['kor', 'eng', 'math']].min().tolist()
print(smin)
score_min=score.fillna(score.min()).copy()
score_min


[74.0, 79.0, 74.0]


sno   kor   eng  math  year
0   1001  88.0  84.0    92  2020
1   1001  74.0  79.0    82  2021
2   1002  84.0  88.0    94  2020
3   1002  74.0  79.0    79  2021
4   1003  92.0  94.0    90  2020
5   1003  86.0  79.0    84  2021
6   1004  88.0  84.0    88  2020
7   1004  79.0  82.0    74  2021
8   1005  90.0  94.0    90  2020
9   1005  84.0  79.0    86  2021
10  1006  88.0  90.0    92  2020
11  1006  74.0  82.0    79  2021
12  1007  92.0  84.0    94  2020
13  1007  74.0  84.0    74  2021

In [15]:
##################
# 혼자 공부한 것 #
##################

# 결과값 저장은 생략

# ------ score 복제 df 만들기

scoretttt = pd.read_csv('./dataset/score.csv')
scoretttt.isna().sum()


# ------ key를 column 이름으로 하는 임시 딕셔너리 만들기(value는 na일 경우 채울 값)

tempd={'kor':9999}


# ------ fillna에 딕셔너리를 넣으면 각 column의 이름과 같은 이름의 'key'가 연결되어 해당 column의 na에 value가 채워진다.

scoretttt.fillna(tempd)


# ------ index를 column 이름으로 하는 임시 시리즈 만들기(값은 na일 경우 채울 값)

temps=pd.Series([6666], index=['eng'])


# ------ fillna에 시리즈를 넣으면 각 column의 이름과 같은 이름의 index가 연결되어 해당 column의 na에 값이 채워진다.

scoretttt.fillna(temps)


# ------ -----응용!!
# ------ 그룹함수 사용하기

scoretttt.min() #그룹함수를 사용하면 결과는 series로 나온다.
scoretttt.fillna(scoretttt.min()) # Na가 각 column의 최소값으로 채워진다.

scoretttt.min().to_dict() # dict로 해도 마찬가지! (굳이..?)
scoretttt.fillna(scoretttt.min().to_dict())

sno   kor   eng  math  year
0   1001  88.0  84.0    92  2020
1   1001  74.0  79.0    82  2021
2   1002  84.0  88.0    94  2020
3   1002  74.0  79.0    79  2021
4   1003  92.0  94.0    90  2020
5   1003  86.0  79.0    84  2021
6   1004  88.0  84.0    88  2020
7   1004  79.0  82.0    74  2021
8   1005  90.0  94.0    90  2020
9   1005  84.0  79.0    86  2021
10  1006  88.0  90.0    92  2020
11  1006  74.0  82.0    79  2021
12  1007  92.0  84.0    94  2020
13  1007  74.0  84.0    74  2021

In [16]:
# 떼거지로 했을 때 inplace=True 하면 제대로 안 먹는다!!

In [17]:
score['kor'].min()
score['eng'].min()

79.0

In [18]:
score1 = score.copy()
#score1['kor'].fillna(50, inplace=True)
#score1['eng'].fillna(50, inplace=True)
score1[  ['kor','eng'] ] = score1[  ['kor','eng'] ].fillna(50)
score1.isna().sum()


sno     0
kor     0
eng     0
math    0
year    0
dtype: int64

In [19]:
# 성적이 없는 학생의 점수는 '해당 과목의 최저 점수'로 채우기
score['kor'].fillna(score['kor'].min(), inplace=True)
score['eng'].fillna(score['eng'].min(), inplace=True)


In [20]:
score.isna().sum()


sno     0
kor     0
eng     0
math    0
year    0
dtype: int64

### 4. kor + eng 평균점수 lang 컬럼 추가

In [55]:
score['lang']=(score['kor'] + score['eng'])/2
# score.head()

# 보너스로 타입 바꾸기
score[['kor','eng', 'lang']]=score[['kor','eng','lang']].astype('int')
score.head()

KeyError: 'kor'

### 5. 모든 수강생의 국,영,수 점수 (년도,학번 정렬)

**** sort_values  
order by sno DESC == ascending=False  
order by sno ASC == (ascending=True) 안 써도 됨.  

In [22]:
# 문제 요지 파악 못 함^^;;;
score.sort_values(by=['year', 'sno'])

sno   kor   eng  math  year
0   1001  88.0  84.0    92  2020
2   1002  84.0  88.0    94  2020
4   1003  92.0  94.0    90  2020
6   1004  88.0  84.0    88  2020
8   1005  90.0  94.0    90  2020
10  1006  88.0  90.0    92  2020
12  1007  92.0  84.0    94  2020
1   1001  74.0  79.0    82  2021
3   1002  74.0  79.0    79  2021
5   1003  86.0  79.0    84  2021
7   1004  79.0  82.0    74  2021
9   1005  84.0  79.0    86  2021
11  1006  74.0  82.0    79  2021
13  1007  74.0  84.0    74  2021

<pre>
아래와 같은 꼴로 출력
sno ename  age gen  kor  eng  math  year  lang
0   1001   송강호   33   m   88   84    92  2020    86
2   1002   황정민   29   m   84   88    94  2020    86
4   1003   김윤석   31   m   92   94    90  2020    93
6   1004   설경구   27   m   88   84    88  2020    86
8   1005   김혜수   32   f   90   94    90  2020    92
10  1006   하정우   23   m   88   90    92  2020    89
12  1007   전지현   26   f   92   84    94  2020    88
1   1001   송강호   33   m   74   79    82  2021    76
3   1002   황정민   29   m   74   79    79  2021    76
5   1003   김윤석   31   m   86   79    84  2021    82
7   1004   설경구   27   m   79   82    74  2021    80
9   1005   김혜수   32   f   84   79    86  2021    81
11  1006   하정우   23   m   74   82    79  2021    78
13  1007   전지현   26   f   74   84    74  2021    79


In [23]:
# 강사님 -> join 하라는 거였다 ^^;;;;;;;

In [24]:
# 조인
# select * from user u, score s where u.sno=s.sno
# select * from user u INNER JOIN score s ON u.sno=s.sno

user_score = pd.merge(user, score, how='inner', on='sno')
user_score

score.sort_values(by=['year', 'sno'])
# sort_values는 inplace, ignore_index 둘 다 false -> 그냥 기본 확인하는 용이라.
# ascending -> True 작은 것부터 정렬, False: Descending

sno   kor   eng  math  year
0   1001  88.0  84.0    92  2020
2   1002  84.0  88.0    94  2020
4   1003  92.0  94.0    90  2020
6   1004  88.0  84.0    88  2020
8   1005  90.0  94.0    90  2020
10  1006  88.0  90.0    92  2020
12  1007  92.0  84.0    94  2020
1   1001  74.0  79.0    82  2021
3   1002  74.0  79.0    79  2021
5   1003  86.0  79.0    84  2021
7   1004  79.0  82.0    74  2021
9   1005  84.0  79.0    86  2021
11  1006  74.0  82.0    79  2021
13  1007  74.0  84.0    74  2021

### 6-1. 남학생, 여학생 평균 나이
### 6-2. 남학생, 여학생 수
### 6-3. 년도별 국영수 최대점수, 최소점수

In [25]:
# 6-1
user.groupby('gen').mean('age')
# 왜 이렇게 하면 sno도 나올까?????

# 강사님
# select avg(age) from user group by gen; 

user.groupby('gen')['age'].mean()

gen
f    29.0
m    28.6
Name: age, dtype: float64

In [26]:
# 6-2
user.value_counts('gen') # Series로 출력됨.

# 강사님
# select count(gen) from user group by gen;

user['gen'].value_counts() # 도 가능
# user.value_counts('gen').to_dict() # dict로 출력

m    5
f    2
Name: gen, dtype: int64

In [27]:
# 6-3
score.groupby('year')[['kor','eng','math']].agg(['max','min'])

# 강사님
# select year min(kor), max(kor) from score group by year;
# select year min(eng), max(eng) from score group by year;
# select year min(math), max(math) from score group by year;

#오라클처럼 따로 안 해도 됨.
# 사실 by= 안 써도 되는데 연습용 ^^
score.groupby(by='year')[['kor','eng','math']].min()
score.groupby(by='year')[['kor','eng','math']].max()

# 합치려면
score.groupby(by='year')[['kor','eng','math']].agg(['min','max'])

kor         eng       math    
       min   max   min   max  min max
year                                 
2020  84.0  92.0  84.0  94.0   88  94
2021  74.0  86.0  79.0  84.0   74  86

### 7. 남학생의 국어 점수 중 가장 높은 점수, 가장 낮은 점수

In [28]:
user_score[user_score['gen']=='m']['kor'].agg(['min','max'])

min    74.0
max    92.0
Name: kor, dtype: float64

### 8. user 정보에 가산점(extpnt) 컬럼 추가 
    남:1 여:0

* 추천 문법  
    <pre>
    <font color='red' size='3'>
    df[컬럼].apply(lambda x : 한줄표현함수) (안 복잡하면!!)  
    
            ->복잡하면 함수로 만들어서 쓰기.  
    
    df.loc[조건(줄),컬럼(칸)] =변경값  

In [29]:
temp=[]
for gen in user['gen'].tolist():
    if gen=='m':
         temp.append(1)
    elif gen=='f':
         temp.append(0)

temp

user['extpnt']=temp

user


sno ename  age gen  extpnt
0  1001   송강호   33   m       1
1  1002   황정민   29   m       1
2  1003   김윤석   31   m       1
3  1004   설경구   27   m       1
4  1005   김혜수   32   f       0
5  1006   하정우   23   m       1
6  1007   전지현   26   f       0

In [30]:
# 또다른 시도

# user[user['gen']=='m']['extpnt']=33

# user22 = pd.read_csv('./dataset/user.csv')

# user22['extpnt']=0
# user22

# user22['extpnt'][user22['gen']=='f']=0
# user22['extpnt'][user22['gen']=='m']=1

# user22

loc 이용 *****************(람다 한 줄 표현 못 하겠으면 이거 써라)

In [31]:
# 강사님
# 강사님 주피터 필기 참고!!!

user = pd.read_csv('./dataset/user.csv')

# 아마 다 이렇게 시도했을듯
#  user[user['gen']=='m']['extpnt']=1 ## 경고!!!!!
## 원본 10 개 -> 그 중 조건 충족하는 5개 -> 이 일부 바꾼 걸 전체에 적용하려 하면 문제(부분부분 잘라서 들어가니까)
# user['extpnt']=1 # 이건 가능 -> 전체에다가 넣으라는 거라(부분부분 아님)

# See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

# loc이나 iloc 이용해서 .__getitem__ 으로 꺼내고 .__setitem__으로 설정
# get set은 좀 귀찮으니 다음과 같은 방법 이용.

# df.loc[조건(줄), 컬럼(칸)]=값 세팅 !!!!
user.loc[user['gen']=='m', 'extpnt']=1
user.loc[user['gen']=='f', 'extpnt']=0
user['extpnt']=user['extpnt'].astype('int') # float 돼서
user



sno ename  age gen  extpnt
0  1001   송강호   33   m       1
1  1002   황정민   29   m       1
2  1003   김윤석   31   m       1
3  1004   설경구   27   m       1
4  1005   김혜수   32   f       0
5  1006   하정우   23   m       1
6  1007   전지현   26   f       0

In [32]:
# 또다른 방법 (루프)

user['extpnt']=1

user = pd.read_csv('./dataset/user.csv')
for j in range(len(user)):
    if user.iloc[j,3]=='m':
        user.iloc[j,4]=1 # 세번째 열이 gen이라, 근데 네 번째 없으니까 미리 만들어놔야.
    else:
        user.iloc[j,4]=0

#비추한다.        
    
    
# cf. for문 얼마나 꺼낼까.
user.shape[0] # .shape하면 튜플로 행개수,열개수 나옴. .shape[0] 하면 행개수
len(user) # 에 range씌우기



IndexError: iloc cannot enlarge its target object

In [33]:
# 인덱스 이용하는 법 ->이것도 비추
user = pd.read_csv('./dataset/user.csv')
user['extpnt']=0

idx_list=user[user['gen']=='m'].index.tolist()
idx_list
user.loc[idx_list, 'extpnt']=1



* 방법(5) 함수를 만들어서 사용
--> 람다 해보자!! (한 줄 for문, 함수 만들기 할 수 있어야)

<pre>
이거 없이 머신러닝 할 수 없다!!!!!!!!!!!******* 무조건 암기
🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕🍕
<font color=red size=3>
df[컬럼11].apply(     lambda 컬럼11  : 적용할함수이름(컬럼11)     )
df[컬럼11].apply(     lambda x  : 적용할함수이름(x)     )

In [34]:
# 함수이름: my_extpnt
# 파라미터: prmgen
#  남성(m): 1 여성(f): 0
# return res <- 남성이면 1 여성이면 0


def my_extpnt(prmgen):
    # res=0 #넣으셨음. (함수 파악하기 좋을듯)
    if prmgen=='m':
        res=1
    elif prmgen=='f':
        res=0
    return res

my_extpnt('f') # ------0
my_extpnt('m') # ------1

# 지금은 파라미터에 f나 m 하나씩 넣음.
# gen 시리즈로 넣고 싶은데 파라미터에 떼거지로 넣는 방법은 없을까

1

In [35]:
# 내부적으로 루프 돌면서 함수에 대입하면서 메모리 공간에 쌓아둠. -> 이걸 다시 넣어줘해야됨.
# 외우는 타이핑!!
# df['gen'] : df['gen']이라는 시리즈를
# apply(lambda gen ~~ : # lambda에 gen으로 받겠다. gen=df['gen'] 이런 느낌
# apply(lambda x ~~ : # lambda에 x로 받겠다. gen=df['gen'] 이런 느낌 이것도 가능
# lambda gen: my_extpnt(gen) : lambda의 gen에 받아온 걸 my_extpnt함수에 gen으로 넣겠다.
# lambda x: my_extpnt(x) :  lambda의 x에 받아온 걸 my_extpnt함수에 x로 넣겠다.


user = pd.read_csv('./dataset/user.csv')
user['extpnt']=user['gen'].apply(lambda x: my_extpnt(x))
user

sno ename  age gen  extpnt
0  1001   송강호   33   m       1
1  1002   황정민   29   m       1
2  1003   김윤석   31   m       1
3  1004   설경구   27   m       1
4  1005   김혜수   32   f       0
5  1006   하정우   23   m       1
6  1007   전지현   26   f       0

한 줄 표현 lambda ************ 제일 강추

In [36]:
# 람다 쓰려면 한 줄 표현 중요.
# res = 1 if prmgen=='m' else 0 

In [37]:
user = pd.read_csv('./dataset/user.csv')
user['extpnt']=user['gen'].apply(lambda x: 1 if x=='m' else 0 )
user

sno ename  age gen  extpnt
0  1001   송강호   33   m       1
1  1002   황정민   29   m       1
2  1003   김윤석   31   m       1
3  1004   설경구   27   m       1
4  1005   김혜수   32   f       0
5  1006   하정우   23   m       1
6  1007   전지현   26   f       0

* age_gubun : 신규 컬럼 생성

<pre>
(1)
20~29이면 2
30~39이면 3
----------
(2)
    flexible 하게
40~49이면 4
..
..


In [38]:
user = pd.read_csv('./dataset/user.csv')

"""
user['age_gugan']
(1)
20~29이면 2
30~39이면 3
----------
(2) flexible 하게
40~49이면 4
"""

# (1)
user['age_gugan']=user['age'].apply(lambda x : 2 if ((x>=20) and (x<30)) else 3 )
user

# (2)
user['age_gugan']=user['age'].apply(lambda x : x//10)
user

sno ename  age gen  age_gugan
0  1001   송강호   33   m          3
1  1002   황정민   29   m          2
2  1003   김윤석   31   m          3
3  1004   설경구   27   m          2
4  1005   김혜수   32   f          3
5  1006   하정우   23   m          2
6  1007   전지현   26   f          2

In [39]:
# 재미로 사람 추가...
user = pd.read_csv('./dataset/user.csv')

new_dict={'sno': [1008,1009,1010,1011,1012,1013],
         'ename': ['김말숙', '김정숙', '김조숙', '김숙숙', '김영숙', '김백숙'],
         'age': [63,45,87,14,70,112],
         'gen': ['f','f','f','f','f','f']}

new_df=pd.DataFrame(new_dict, columns=user.columns)
new_df

user=pd.concat([user, new_df], ignore_index=True)

# (1)
user['age_gugan']=user['age'].apply(lambda x : 2 if ((x>=20) and (x<30)) else 3 )
user

# (2)
user['age_gugan']=user['age'].apply(lambda x : x//10)
user



sno ename  age gen  age_gugan
0   1001   송강호   33   m          3
1   1002   황정민   29   m          2
2   1003   김윤석   31   m          3
3   1004   설경구   27   m          2
4   1005   김혜수   32   f          3
5   1006   하정우   23   m          2
6   1007   전지현   26   f          2
7   1008   김말숙   63   f          6
8   1009   김정숙   45   f          4
9   1010   김조숙   87   f          8
10  1011   김숙숙   14   f          1
11  1012   김영숙   70   f          7
12  1013   김백숙  112   f         11

In [40]:
# 강사님
# if age in range(20, 30): # 20,21,...29 이렇게도 가능.

# 복잡할 땐 따로 함수 만들어라!!
user = pd.read_csv('./dataset/user.csv')

def myfunc(age):
    res=0
    if age in range(20,30):
        res=2
    elif age in range(30,40):
        res=3
    elif age in range(40,50):
        res=4
    return res
user['age_gubun']=user['age'].apply(lambda x : myfunc(x))
user

sno ename  age gen  age_gubun
0  1001   송강호   33   m          3
1  1002   황정민   29   m          2
2  1003   김윤석   31   m          3
3  1004   설경구   27   m          2
4  1005   김혜수   32   f          3
5  1006   하정우   23   m          2
6  1007   전지현   26   f          2

In [41]:
user = pd.read_csv('./dataset/user.csv')

user['age_gubun']=user['age'].apply(lambda x : x//10)
user

sno ename  age gen  age_gubun
0  1001   송강호   33   m          3
1  1002   황정민   29   m          2
2  1003   김윤석   31   m          3
3  1004   설경구   27   m          2
4  1005   김혜수   32   f          3
5  1006   하정우   23   m          2
6  1007   전지현   26   f          2

(1) 국어 점수 결측 : 50  
(2) 국어, 영어, 수학 점수 결측 : 50

In [42]:
score = pd.read_csv('./dataset/score.csv')

# 시도 실패!!!
# score['kor']=score['kor'].apply(lambda x : 50 if (bool(x)==False))
# score['kor']=score['kor'].apply(lambda x : 50 if bool(x)==False else x)
# score['kor']=score['kor'].apply(lambda x : 50 if x=='nan' else x)
# score['kor']=score['kor'].apply(lambda x : 50 if x==np.nan else x)

# 최상현
# 프레임의 컬럼에 담긴 값을 넘겨서 비교하는 경우: np.isnan(k)
score['kor2']=score['kor'].apply(lambda x: 50 if np.isnan(x) else x)

score

# 프레임 통째로 넘겨서 비교하는 경우(axis 반드시 표기!!!!)
# 프레임 통째로 가져올 때는 axis 꼭 넣어줘야함!!!!!!!!! 방향 몰라~
score['kor2']=score.apply(lambda dfx: 50 if pd.isnull(dfx['kor']) else dfx['kor'], axis=1)

score

# 문법 매번 바뀌니까 뭐가 되는지 봐라

sno   kor   eng  math  year  kor2
0   1001  88.0  84.0    92  2020  88.0
1   1001   NaN  79.0    82  2021  50.0
2   1002  84.0  88.0    94  2020  84.0
3   1002  74.0   NaN    79  2021  74.0
4   1003  92.0  94.0    90  2020  92.0
5   1003  86.0   NaN    84  2021  86.0
6   1004  88.0  84.0    88  2020  88.0
7   1004  79.0  82.0    74  2021  79.0
8   1005  90.0  94.0    90  2020  90.0
9   1005  84.0   NaN    86  2021  84.0
10  1006  88.0  90.0    92  2020  88.0
11  1006  74.0  82.0    79  2021  74.0
12  1007  92.0  84.0    94  2020  92.0
13  1007   NaN  84.0    74  2021  50.0

In [43]:
# 허강호
score[['k','e','m']] = score[['kor','eng','math']].apply(lambda F : F.fillna(50))
score

# 사실 lambda 안 쓰고 fillna 쓰는 게 제일 간단

sno   kor   eng  math  year  kor2     k     e   m
0   1001  88.0  84.0    92  2020  88.0  88.0  84.0  92
1   1001   NaN  79.0    82  2021  50.0  50.0  79.0  82
2   1002  84.0  88.0    94  2020  84.0  84.0  88.0  94
3   1002  74.0   NaN    79  2021  74.0  74.0  50.0  79
4   1003  92.0  94.0    90  2020  92.0  92.0  94.0  90
5   1003  86.0   NaN    84  2021  86.0  86.0  50.0  84
6   1004  88.0  84.0    88  2020  88.0  88.0  84.0  88
7   1004  79.0  82.0    74  2021  79.0  79.0  82.0  74
8   1005  90.0  94.0    90  2020  90.0  90.0  94.0  90
9   1005  84.0   NaN    86  2021  84.0  84.0  50.0  86
10  1006  88.0  90.0    92  2020  88.0  88.0  90.0  92
11  1006  74.0  82.0    79  2021  74.0  74.0  82.0  79
12  1007  92.0  84.0    94  2020  92.0  92.0  84.0  94
13  1007   NaN  84.0    74  2021  50.0  50.0  84.0  74

### Group by 하면 apply 또 안 된다.
group by 하면 데이터 축약되는데 여기에 apply 하면 이 축약버전에 대입된다. -> Transform 써야됨.  
그냥 쓸 때는 apply나 transform이나 같은데 group by하고 Transform에 lambda 쓰면 방법 또 다르다.  
버전마다 문법마다 다 다르니까 그냥 잘 된 거 저장해놓고 복붙해라

## 람다식 적용 시 transform() vs apply() ( 나중에)

### 9-1. kor, eng 컬럼 삭제
### 9-2. 인덱스 값이 0,2인 row 삭제

In [54]:
score = pd.read_csv('./dataset/score.csv')

# 9-1
score.drop(['kor','eng'], axis=1, inplace=True)

# 9-2
# 인덱스 '값'이다! '번째'아니다!!!*********************************
score.drop([0,2], axis=0, inplace=True) # axis=0 생략 가능!

score

sno  math  year
1   1001    82  2021
3   1002    79  2021
4   1003    90  2020
5   1003    84  2021
6   1004    88  2020
7   1004    74  2021
8   1005    90  2020
9   1005    86  2021
10  1006    92  2020
11  1006    79  2021
12  1007    94  2020
13  1007    74  2021

### 10.1008, 이정재35 m 추가

In [84]:
score = pd.read_csv('./dataset/score.csv')
user = pd.read_csv('./dataset/user.csv')

# 10-1 (1) (리스트 -> Series & append)
temp=[1008, '이정재', 35, 'm']
temps=pd.Series(temp, index=user.columns)
user=user.append(temps, ignore_index=True)

# 10-1 (2) (리스트 -> df & concat)
# temp2=[1008, '이정재', 35, 'm']
# tempdf=pd.DataFrame([temp2],columns=user.columns)
# user=pd.concat([user,  tempdf], ignore_index=True)

# 10-2 -> 11
user_score=pd.merge(user, score[score['year']==2021], how='left' ,on='sno')
user_score

sno ename  age gen   kor   eng  math    year
0  1001   송강호   33   m   NaN  79.0  82.0  2021.0
1  1002   황정민   29   m  74.0   NaN  79.0  2021.0
2  1003   김윤석   31   m  86.0   NaN  84.0  2021.0
3  1004   설경구   27   m  79.0  82.0  74.0  2021.0
4  1005   김혜수   32   f  84.0   NaN  86.0  2021.0
5  1006   하정우   23   m  74.0  82.0  79.0  2021.0
6  1007   전지현   26   f   NaN  84.0  74.0  2021.0
7  1008   이정재   35   m   NaN   NaN   NaN     NaN

In [88]:
# 강사님
# 이 문법 어렵다 
score = pd.read_csv('./dataset/score.csv')
user = pd.read_csv('./dataset/user.csv')

add_list=[1008, '이정재', 35, 'm']
# user.append(DF 또는 Series만 가능 )

s=pd.Series(add_list, index=user.columns)
user.append(s,ignore_index=True)
user


sno ename  age gen
0  1001   송강호   33   m
1  1002   황정민   29   m
2  1003   김윤석   31   m
3  1004   설경구   27   m
4  1005   김혜수   32   f
5  1006   하정우   23   m
6  1007   전지현   26   f

In [96]:
# 한 건 넣을 때 쉬운 방법
# 보통은 한 건 안 넣고 여러 건 넣는다. -> 일반적인 건 frame에 frame 합치기

user = pd.read_csv('./dataset/user.csv')

# loc에서 없는 인덱스 지정하면 추가됨.
user.loc[7]=[1008,'이정재',35,'m']

# 7 하드코딩 안 하려면
user.index.max()+1 # 강사님은 values 넣는 거 알려주셨는데 안 넣어도 된다.

user.index.values # 하면 array 된다.
user.index.values.max()+1

# 어려우면 직접 뽑아와도 된다.
user.iloc[-1] # series로 마지막 거 가져옴. 이 series의 name이 index 이름.
user.iloc[-1].name+1 # 마지막 번호 뽑아오기


user.loc[user.index.max()+1]=[1008,'이정재',35,'m']
user



sno ename  age gen
0  1001   송강호   33   m
1  1002   황정민   29   m
2  1003   김윤석   31   m
3  1004   설경구   27   m
4  1005   김혜수   32   f
5  1006   하정우   23   m
6  1007   전지현   26   f
7  1008   이정재   35   m
8  1008   이정재   35   m

In [94]:
#df+df

add_list=[1008, '이정재', 35, 'm']
add_df=pd.DataFrame([temp2],columns=user.columns)
user=pd.concat([user,  add_df], ignore_index=True)

### 11 2021년 모든 수강생 국영수 점수 출력(성적 없어도 출력)

* pd.concat([df1, df2]) : 합치기 (axis 주의)/ axis=1로 좌우 하면 full outer join이라 비추  
* pd.merge(df1, df2) : join 

In [95]:
# 11 <-10-2
user_score=pd.merge(user, score[score['year']==2021], how='left' ,on='sno')
user_score

sno ename  age gen   kor   eng  math    year
0  1001   송강호   33   m   NaN  79.0  82.0  2021.0
1  1002   황정민   29   m  74.0   NaN  79.0  2021.0
2  1003   김윤석   31   m  86.0   NaN  84.0  2021.0
3  1004   설경구   27   m  79.0  82.0  74.0  2021.0
4  1005   김혜수   32   f  84.0   NaN  86.0  2021.0
5  1006   하정우   23   m  74.0  82.0  79.0  2021.0
6  1007   전지현   26   f   NaN  84.0  74.0  2021.0
7  1008   이정재   35   m   NaN   NaN   NaN     NaN
8  1008   이정재   35   m   NaN   NaN   NaN     NaN
9  1008   이정재   35   m   NaN   NaN   NaN     NaN

# 내일(220112) 할 것 ->
### 그룹값 변경 : transform() vs. apply()
### 정규화 -> 양식 맞춰서 쓰도록
### 시각화